### ISSUE TITLE
padding mask and attention mask

### ISSUE THREAD

**Issue by zackchen-lb** [tatp22/linformer-pytorch#11]
Hi there,

You've done a great job and thanks for the sharing. I'm wondering how you deal with the masking stuff in Linformer since the attention shape, key and value shape have now changed to (n, k) instead of (n, n). I didn't find these in the code. Thanks for your time!

**Comment by tatp22** [comment]
Hi @ZEKAICHEN,

I'm not sure I completely understand the question, so I will break it up into two parts, hopefully one of them will answer your question:

If you are looking for a way to pad your inputs, check out my `Padder` class, which can be found [here](https://github.com/tatp22/linformer-pytorch#padder). Unfortunately, I have not have time to test it on the `LinfromerLM` module, and I will be doing that in the upcoming days, hopefully.

As for masking, that's the next thing that I want to get done off my bucket list. Look for it in a coming update.

If this is not what you wanted answered, let me know! I'll try and help as much as I can

**Comment by zackchen-lb** [comment]
Hi @tatp22 ,

Thanks for the feedback. I was exactly talking about the masking. Thanks for letting me know.

**Comment by tatp22** [comment]
Hi @ZEKAICHEN,

Actually, what you said is kinda a dilemma, since I actually cannot mask it in the traditional sense (the traditional sense being exactly what you said before). Therefore, my idea was to just mask the Q, K, and V matrices directly, similar to what is done [here](https://github.com/lucidrains/linear-attention-transformer/blob/0cca90fd77a171fb5b8d80a143f22f8a7543c170/linear_attention_transformer/linear_attention_transformer.py#L176). I think this is the best way to do it, and I will mask it like this; let me know if you think this is good, and if not, I am open up for ideas.

In theory, I think it should work the same (I can come up with a quick proof or something, maybe), and this is the only logical way that I see that we can pad the inputs. I'll leave this issue open until I make the pull request.

**Comment by tatp22** [comment]
I added more information in the PR. If it looks good, let me know :+1:, otherwise I will just merge it, since it really doesn't affect the rest of the code too much

**Comment by zackchen-lb** [comment]
I agree with masking projection matrices directly and thanks for the updates. But there could be some other solutions maybe? I'm also working on this but based on fairseq. Thanks again.

**Comment by tatp22** [comment]
Not sure what else to think of :shrug: This is the best solution I could come up with; I even emailed the authors of the paper about this, but no response, unfortunately...

If no one says anything else, I will merge this soon.

**Comment by tatp22** [comment]
Merging

**Comment by tatp22** [comment]
Actually, sorry to keep on closing and opening this up again, but another possibility to mask would be to mask in the traditional sense, but instead of the entire (n,n) masking matrix, we can just make an (n,k) masking matrix, and then make it upper triangular as well, but just have the second dimension be cut off. This can work as well, let me know what you think.

**Comment by tatp22** [comment]
Hey, I added another upper triangular padding mask option for causal sequences, let me know what you think. It's up on the latest commit, and if the dims are `(n,k)`, it simply cuts off the right part of the matrix.

I'll close this issue now, since I added both options and now users can use whichever one they want.

**Comment by zackchen-lb** [comment]
> Hey, I added another upper triangular padding mask option for causal sequences, let me know what you think. It's up on the latest commit, and if the dims are `(n,k)`, it simply cuts off the right part of the matrix.
> 
> I'll close this issue now, since I added both options and now users can use whichever one they want.

Thanks for the updates. Wonderful solution.
